### ISSUE TITLE
CentOS 7: Cookbook doesn't upgrade pdns-recursor previously installed from EPEL Repo

### ISSUE THREAD

**Issue by MattMencel** [dnsimple/chef-pdns#65]
If pdns-recursor has previously been installed from the default CentOS repository, the pdns cookbook does not attempt to upgrade it.

**Comment by MattMencel** [comment]
Did some testing of this based on the work I did in #64.

If pdns-recursor was previously installed from the EPEL repo, the pdns_recursor_install resource doesn't attempt to upgrade it.

>  * pdns_recursor_install_rhel[my_recursor] action install
>     * yum_package[epel-release] action install (skipped due to only_if)
>     * yum_repository[powerdns-rec-40] action create
>       * template[/etc/yum.repos.d/powerdns-rec-40.repo] action create
>         - create new file /etc/yum.repos.d/powerdns-rec-40.repo
>         - update content in file /etc/yum.repos.d/powerdns-rec-40.repo from none to 1085a6
>         --- /etc/yum.repos.d/powerdns-rec-40.repo	2017-08-08 14:43:42.699241966 +0000
>         +++ /etc/yum.repos.d/.chef-powerdns-rec-4020170808-630-eu5xgv.repo	2017-08-08 14:43:42.699241966 +0000
>         @@ -1 +1,13 @@
>         +# This file was generated by Chef
>         +# Do NOT modify this file by hand.
>         +
>         +[powerdns-rec-40]
>         +name=PowerDNS repository for PowerDNS Recursor - version 4.0.X
>         +baseurl=http://repo.powerdns.com/centos/$basearch/$releasever/rec-40
>         +enabled=1
>         +fastestmirror_enabled=0
>         +gpgcheck=1
>         +gpgkey=https://repo.powerdns.com/FD380FBB-pub.asc
>         +includepkgs=pdns*
>         +priority=90
>         - change mode from '' to '0644'
>       * execute[yum clean metadata powerdns-rec-40] action run
>         - execute yum clean metadata --disablerepo=* --enablerepo=powerdns-rec-40
>       * execute[yum-makecache-powerdns-rec-40] action run
>         - execute yum -q -y makecache --disablerepo=* --enablerepo=powerdns-rec-40
>       * ruby_block[yum-cache-reload-powerdns-rec-40] action create
>         - execute the ruby block yum-cache-reload-powerdns-rec-40
>       * execute[yum clean metadata powerdns-rec-40] action nothing (skipped due to action :nothing)
>       * execute[yum-makecache-powerdns-rec-40] action nothing (skipped due to action :nothing)
>       * ruby_block[yum-cache-reload-powerdns-rec-40] action nothing (skipped due to action :nothing)
> 
>     * yum_repository[powerdns-rec-40-debuginfo] action create
>       * template[/etc/yum.repos.d/powerdns-rec-40-debuginfo.repo] action create
>         - create new file /etc/yum.repos.d/powerdns-rec-40-debuginfo.repo
>         - update content in file /etc/yum.repos.d/powerdns-rec-40-debuginfo.repo from none to 823889
>         --- /etc/yum.repos.d/powerdns-rec-40-debuginfo.repo	2017-08-08 14:43:43.919286285 +0000
>         +++ /etc/yum.repos.d/.chef-powerdns-rec-40-debuginfo20170808-630-b0lin2.repo	2017-08-08 14:43:43.919286285 +0000
>         @@ -1 +1,13 @@
>         +# This file was generated by Chef
>         +# Do NOT modify this file by hand.
>         +
>         +[powerdns-rec-40-debuginfo]
>         +name=PowerDNS repository for PowerDNS Recursor - version 4.0.X debug symbols
>         +baseurl=http://repo.powerdns.com/centos/$basearch/$releasever/rec-40/debug
>         +enabled=1
>         +fastestmirror_enabled=0
>         +gpgcheck=1
>         +gpgkey=https://repo.powerdns.com/FD380FBB-pub.asc
>         +includepkgs=pdns*
>         +priority=90
>         - change mode from '' to '0644'
>       * execute[yum clean metadata powerdns-rec-40-debuginfo] action run
>         - execute yum clean metadata --disablerepo=* --enablerepo=powerdns-rec-40-debuginfo
>       * execute[yum-makecache-powerdns-rec-40-debuginfo] action run
>         - execute yum -q -y makecache --disablerepo=* --enablerepo=powerdns-rec-40-debuginfo
>       * ruby_block[yum-cache-reload-powerdns-rec-40-debuginfo] action create
>         - execute the ruby block yum-cache-reload-powerdns-rec-40-debuginfo
>       * execute[yum clean metadata powerdns-rec-40-debuginfo] action nothing (skipped due to action :nothing)
>       * execute[yum-makecache-powerdns-rec-40-debuginfo] action nothing (skipped due to action :nothing)
>       * ruby_block[yum-cache-reload-powerdns-rec-40-debuginfo] action nothing (skipped due to action :nothing)
> 
>     * yum_package[pdns-recursor] action install (up to date)

**Comment by MattMencel** [comment]
Maybe just provide an upgrade action on the pdns_recursor_install_rhel resource?  I can put a PR in unless you'd prefer not to support that action.  It could be done in a wrapper cookbook instead.

**Comment by onlyhavecans** [comment]
I think an upgrade action is a good idea. My only concern is how well the powerdns repo can upgrade the package from EPEL repo. If you have the idea for a PR in mind I think it would be very welcome

**Comment by MattMencel** [comment]
I upgraded the EPEL package to the powerdns repo package in place manually with 'yum update', so it seems pretty straight forward.

Where I ran into a problem was with the systemd_service_drop_in in my wrapper cookbook.  Because it had overridden the default config, after the new service is installed it wouldn't restart the service.  To work around that in the upgrade action I'll probably need to stop the service first, upgrade, then restart the service.

I'll see about getting a PR in tomorrow.
